cpg-language-llvm/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/CompressLLVMPass.kt

```diff
@@ -72,113 +72,99 @@ class CompressLLVMPass(ctx: TranslationContext) : ComponentPass(ctx) {
-                    }
-                    (node.statement as Block).statements = newStatements
+                    handleSwitchStatement(node, gotosToReplace)
                 }
                 is TryStatement -> {
                     handleTryStatement(node)
                 }
                 is Block -> {
-                    // Get the last statement in a Block and replace a goto statement
-                    // iff it is the only one jumping to the target
-                    val goto = node.statements.lastOrNull()
-                    if (
-                        goto != null &&
-                            goto in gotosToReplace &&
-                            node !in
-                                SubgraphWalker.flattenAST(
-                                    (goto as GotoStatement).targetLabel?.subStatement
-                                )
-                    ) {
-                        val subStatement = goto.targetLabel?.subStatement
-                        val newStatements = node.statements.dropLast(1).toMutableList()
-                        newStatements.addAll((subStatement as Block).statements)
-                        node.statements = newStatements
-                    }
+                    handleBlock(node, gotosToReplace)
                 }
             }
         }
     }
 
+    /**
+     * Get the last statement in a Block and replace a goto statement iff it is the only one jumping
+     * to the target
+     */
+    private fun handleBlock(node: Block, gotosToReplace: List<GotoStatement>) {
+        val goto = node.statements.lastOrNull() as? GotoStatement ?: return
+        val gotoSubstatement = goto.targetLabel?.subStatement as? Block ?: return
+        if (goto in gotosToReplace && node !in gotoSubstatement.allChildren<Block>()) {
+            val newStatements = node.statements.dropLast(1).toMutableList()
+            newStatements.addAll(gotoSubstatement.statements)
+            node.statements = newStatements
+        }
+    }
+
+    /**
+     * Iterates over all statements in a body of the switch/case and replace a goto statement if it
+     * is the only one jumping to the target
+     */
+    private fun handleSwitchStatement(node: SwitchStatement, gotosToReplace: List<GotoStatement>) {
+        val caseBodyStatements = node.statement as? Block ?: return
+        val newStatements = caseBodyStatements.statements.toMutableList()
+        for (i in 0 until newStatements.size) {
+            val subStatement = (newStatements[i] as? GotoStatement)?.targetLabel?.subStatement
+            if (
+                newStatements[i] in gotosToReplace &&
+                    newStatements[i] !in (subStatement?.astChildren ?: listOf())
+            ) {
+                subStatement?.let { newStatements[i] = it }
+            }
+        }
+        caseBodyStatements.statements = newStatements
+    }
+
+    /**
+     * Replace the then-statement and else-statement with the basic block it jumps to iff we found
```

Comment:
typo, also a couple of more of those

Reply:
It was intentional: https://en.wikipedia.org/wiki/If_and_only_if

Reply:
ah, but not all "iffs" in this file are intentional 😆